File: musica/grid.cpp
// Copyright (C) 2023-2025 National Center for Atmospheric Research
// SPDX-License-Identifier: Apache-2.0
//
// This file defines the Python bindings for the TUV-x Grid class in the musica library.
#include "binding_common.hpp"

#include <musica/tuvx/grid.hpp>

#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

namespace py = pybind11;

void bind_tuvx_grid(py::module_ &grid)
{
  py::class_<musica::Grid>(grid, "_Grid")
      .def(py::init(
          [](const py::kwargs &kwargs)
          {
            if (!kwargs.contains("name"))
              throw py::value_error("Missing required argument: name");
            if (!kwargs.contains("units"))
              throw py::value_error("Missing required argument: units");
            if (!kwargs.contains("num_sections"))
              throw py::value_error("Missing required argument: num_sections");
            if (!py::isinstance<py::str>(kwargs["name"]))
              throw py::value_error("Argument 'name' must be a string");
            if (!py::isinstance<py::str>(kwargs["units"]))
              throw py::value_error("Argument 'units' must be a string");
            if (!py::isinstance<py::int_>(kwargs["num_sections"]))
              throw py::value_error("Argument 'num_sections' must be an integer");
            if (kwargs["num_sections"].cast<std::size_t>() <= 0)
              throw py::value_error("Argument 'num_sections' must be greater than 0");
            std::string name = kwargs["name"].cast<std::string>();
            std::string units = kwargs["units"].cast<std::string>();
            std::size_t num_sections = kwargs["num_sections"].cast<std::size_t>();
            musica::Error error;
            auto grid_instance = new musica::Grid(name.c_str(), units.c_str(), num_sections, &error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error creating grid: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            return grid_instance;
          }))
      .def("__del__", [](musica::Grid &grid) {})
      .def_property_readonly(
          "name",
          [](musica::Grid &self)
          {
            musica::Error error;
            std::string name = self.GetName(&error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting grid name: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            musica::DeleteError(&error);
            return name;
          },
          "The name of the grid")
      .def_property_readonly(
          "units",
          [](musica::Grid &self)
          {
            musica::Error error;
            std::string units = self.GetUnits(&error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting grid units: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            musica::DeleteError(&error);
            return units;
          },
          "The units of the grid")
      .def_property_readonly(
          "num_sections",
          [](musica::Grid &self)
          {
            musica::Error error;
            std::size_t num_sections = self.GetNumberOfSections(&error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting number of grid sections: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            return num_sections;
          },
          "The number of sections in the grid")
      .def_property(
          "edges",
          // Getter - converts C++ array to numpy array
          [](musica::Grid &self)
          {
            musica::Error error;
            size_t size = self.GetNumberOfSections(&error) + 1;
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting number of grid sections: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            auto result = py::array_t<double>(size);
            py::buffer_info buf = result.request();
            double *ptr = static_cast<double *>(buf.ptr);
            self.GetEdges(ptr, size, &error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting grid edges: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            return result;
          },
          // Setter - converts numpy array to C++ array
          [](musica::Grid &self, py::array_t<double, py::array::c_style | py::array::forcecast> array)
          {
            py::buffer_info buf = array.request();
            musica::Error error;
            size_t size = self.GetNumberOfSections(&error) + 1;
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting number of grid sections: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            if (buf.ndim != 1)
            {
              throw py::value_error("Number of dimensions must be one");
            }
            if (static_cast<size_t>(buf.size) != size)
            {
              throw py::value_error("Array size must be num_sections + 1");
            }
            double *ptr = static_cast<double *>(buf.ptr);
            self.SetEdges(ptr, size, &error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error setting grid edges: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
          },
          "Grid edges array of length num_sections + 1")
      .def_property(
          "midpoints",
          // Getter - converts C++ array to numpy array
          [](musica::Grid &self)
          {
            musica::Error error;
            size_t size = self.GetNumberOfSections(&error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting number of grid sections: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            auto result = py::array_t<double>(size);
            py::buffer_info buf = result.request();
            double *ptr = static_cast<double *>(buf.ptr);
            self.GetMidpoints(ptr, size, &error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting grid midpoints: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            return result;
          },
          // Setter - converts numpy array to C++ array
          [](musica::Grid &self, py::array_t<double, py::array::c_style | py::array::forcecast> array)
          {
            py::buffer_info buf = array.request();
            musica::Error error;
            size_t size = self.GetNumberOfSections(&error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error getting number of grid sections: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
            if (buf.ndim != 1)
            {
              throw py::value_error("Number of dimensions must be one");
            }
            if (static_cast<size_t>(buf.size) != size)
            {
              throw py::value_error("Array size must be num_sections");
            }
            double *ptr = static_cast<double *>(buf.ptr);
            self.SetMidpoints(ptr, size, &error);
            if (!musica::IsSuccess(error))
            {
              std::string message = "Error setting grid midpoints: " + std::string(error.message_.value_);
              musica::DeleteError(&error);
              throw py::value_error(message);
            }
          },
          "Grid midpoints array of length num_sections");
}

File: musica/grid.py
# Copyright (C) 2023-2025 National Center for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0
"""
TUV-x Grid class.

This module provides a class for defining grids on which TUV-x profiles exist.
Typically, this would be used to define vertical and wavelength grids.

Note: TUV-x is only available on macOS and Linux platforms.
"""

from typing import Dict, Optional
import numpy as np
from . import backend

_backend = backend.get_backend()

Grid = _backend._tuvx._Grid if backend.tuvx_available() else None

if backend.tuvx_available():
    original_init = Grid.__init__

    def __init__(self, *, name: str, units: str,
                 num_sections: Optional[int] = None,
                 edges: Optional[np.ndarray] = None,
                 midpoints: Optional[np.ndarray] = None,
                 **kwargs):
        """Initialize a Grid instance. Note that at least one of num_sections, edges, or midpoints
        must be provided.

        Args:
            name: Name of the grid
            units: Units of the grid values
            num_sections: Optional number of grid sections
            edges: Optional array of edge values (length num_sections + 1)
            midpoints: Optional array of midpoint values (length num_sections)
            **kwargs: Additional arguments passed to the C++ constructor
        """
        if (num_sections is None and edges is None and midpoints is None):
            raise ValueError("At least one of num_sections, edges, or midpoints must be provided.")
        if (num_sections is None):
            if (edges is not None):
                num_sections = len(edges) - 1
            elif (midpoints is not None):
                num_sections = len(midpoints)
        # Call the original C++ constructor correctly
        original_init(self, name=name, units=units, num_sections=num_sections, **kwargs)

        # Set edges or midpoints if provided
        if edges is not None:
            self.edges = edges
        if midpoints is not None:
            self.midpoints = midpoints

    Grid.__init__ = __init__

    def __str__(self):
        """User-friendly string representation."""
        return f"Grid(name={self.name}, units={self.units}, num_sections={self.num_sections})"

    Grid.__str__ = __str__

    def __repr__(self):
        """Detailed string representation for debugging."""
        return (f"Grid(name={self.name}, units={self.units}, num_sections={self.num_sections}, "
                f"edges={self.edges}, midpoints={self.midpoints})")

    Grid.__repr__ = __repr__

    def __len__(self):
        """Return the number of sections in the grid."""
        return self.num_sections

    Grid.__len__ = __len__

    def __eq__(self, other):
        """Check equality with another Grid instance."""
        if not isinstance(other, Grid):
            return NotImplemented
        return (self.name == other.name and
                self.units == other.units and
                self.num_sections == other.num_sections and
                np.array_equal(self.edges, other.edges) and
                np.array_equal(self.midpoints, other.midpoints))

    Grid.__eq__ = __eq__

    def __bool__(self):
        """Return True if the grid has sections."""
        return self.num_sections > 0

    Grid.__bool__ = __bool__

    def __contains__(self, value):
        """Check if a value is within the grid bounds."""
        return self.edges[0] <= value <= self.edges[-1]

    Grid.__contains__ = __contains__
